File: python/review/binary_search.py
def binary_search(elements, target):
    def bs(left, right):
        mid = (left + right) // 2
        
        if target == elements[mid]:
            return mid
        elif left > right:
            return -1
        
        if target < elements[mid]:
            return bs(left, mid-1)
        else:
            return bs(mid+1, right)

    return bs(0,len(elements)-1)
        

File: typescript/experiments/binary-search.ts
export function binarySearch(nums: number[], target: number) {
  function bs(start: number, end: number): number {
    const mid = Math.floor((start + end) / 2);
    if (target === nums[mid]) return mid;
    if (start >= end) return -1;
    if (target < nums[mid]) return bs(start, mid - 1);
    else return bs(mid + 1, end);
  }
  return bs(0, nums.length - 1);
}
